### PR TITLE
Add an API to request access to another user with their sharing code

### PR DESCRIPTION
**Pull request by StevePole**
This is how I was thinking to approach the access request.  I haven't implemented the persistence layer as I imagine we're going to persist this through the authorisation system, rather than as a standard CRUD resource.

### CODE REVIEW

File: src/api/routes/accessRequest.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { Request, Response } from 'express';
+import AsyncRouter from '../AsyncRouter';
+import { requestAccess } from '../../application/service';
+import { User } from '../../domain/model/accessRequest/AccessRequest';
+
+export default () => {
+  const route = new AsyncRouter();
+
+  route.post(
+    '/users/:id/access-requests',
```
Comment (kstam) [inline src/api/routes/accessRequest.ts]: It seems like here we are posting an `access-request` to a different userId that doesn't belong to me. Given we already have an approval (sharingCode) this doesn't seem right to me.

I would instead do the following
```
POST /users/:id/access-passes
{ sharingCode: <UUID> }

Where `id` is the user's ID. This way we can also have a GET request that can give us a list of all the users to whom we still have an active pass.

GET /users/:id/access-passes
[ { userId: 10 }. { userId: 20 }, ...]
```

Then we can also implement access-requests (again under the same user's ID) that would not necessarily need a user's presense. Instead we can send a notification to the other user. (Not sure if that will ever be needed)

File: src/api/routes/accessPass.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        const accessPass = await createAccessPass.withSharingCode(code, id);
+        res
+          .json({
+            userId: accessPass.subjectUserId.value,
+            expiryTime: accessPass.expirationTime(),
+          })
+          .status(200);
+      } catch (error) {
+        if (error instanceof AccessPassFailedError) {
+          res.status(403).json({
```
Comment (kstam) [inline src/api/routes/accessPass.ts]: Here it is probably better to throw an ApiError and let it be handled by the general error handling.
```
throw new ApiError(403, error.failureReason.toString());
```

File: src/application/service/CreateAccessPass.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  constructor(
+    private accessPassRepository: AccessPassRepository,
+    private sharingCodeRepository: SharingCodeRepository
+  ) {}
+
+  public async withSharingCode(
+    code: string,
+    userId: string
+  ): Promise<AccessPass> {
+    const sharingCode = await sharingCodeRepository.findByCode(code);
```
Comment (kstam) [inline src/application/service/CreateAccessPass.ts]: here we should also check if the sharing code is not found?

Comment (StevePole) [reply src/application/service/CreateAccessPass.ts]: Yes, since done in my latest 👍

File: src/api/routes/accessPass.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+export default () => {
+  const route = new AsyncRouter();
+
+  route.post(
+    '/users/:id/access-passes',
+    async (req: AuthenticatedRequest, res: Response) => {
+      const { id } = req.params;
+      const { code } = req.body;
+
+      const user = await getUser.byId(id);
```
Comment (kstam) [inline src/api/routes/accessPass.ts]: Do we need this part here? Can't we just use the `accessManager.isLoggedInAsUser(new UserId(id))`?

If it returns true then for sure the user exists. if it returns false we anyway would send 404.

Comment (StevePole) [reply src/api/routes/accessPass.ts]: I think I started this before accessManager existed, but I can switch to it 👍

File: src/infrastructure/persistence/PsqlAccessPassRepository.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  async save(accessPass: AccessPass) {
+    return await this.db
+      .raw(
+        `
+      insert into "${ACCESS_PASS_TABLE_NAME}" (id, actor_user_id, subject_user_id, creation_time)
+      values (:id, :actor_user_id, :subject_user_id, :creation_time)
+      on conflict do nothing
+    `,
+        // TODO something smarter on conflict?
```
Comment (kstam) [inline src/infrastructure/persistence/PsqlAccessPassRepository.ts]: Is there any point where we might want to save an access pass twice? If not I would completely remove the `on conflict` bit. This way the database will fail if a pass with the same ID is attempted to be saved

**Pull request merged**